lib/transport/transport-udp-socket.c

```diff
@@ -0,0 +1,169 @@
+          struct sockaddr *sa = (struct sockaddr *) CMSG_DATA(cmsg);
+
+          log_transport_aux_data_set_local_addr_ref(aux, g_sockaddr_new(sa, cmsg->cmsg_len - cmsg_header_len));
+          break;
+        }
+#else
+      if (cmsg->cmsg_level == IPPROTO_IP && cmsg->cmsg_type == IP_PKTINFO)
+        {
+          struct sockaddr_in sin;
+          struct in_pktinfo *inpkt = (struct in_pktinfo *) CMSG_DATA(cmsg);
+
+          g_assert(self->local_addr->sa.sa_family == AF_INET);
+          sin = *(struct sockaddr_in *) &self->local_addr->sa;
+          sin.sin_addr = inpkt->ipi_addr;
+          log_transport_aux_data_set_local_addr_ref(aux, g_sockaddr_new((struct sockaddr *) &sin, sizeof(sin)));
+          break;
+        }
+#endif
+      if (cmsg->cmsg_level == IPPROTO_IPV6 && cmsg->cmsg_type == IPV6_PKTINFO)
+        {
+          struct sockaddr_in6 sin6;
+          struct in6_pktinfo *in6pkt = (struct in6_pktinfo *) CMSG_DATA(cmsg);
+
+          g_assert(self->local_addr->sa.sa_family == AF_INET6);
+          sin6 = *(struct sockaddr_in6 *) &self->local_addr->sa;
+          sin6.sin6_addr = in6pkt->ipi6_addr;
+          log_transport_aux_data_set_local_addr_ref(aux, g_sockaddr_new((struct sockaddr *) &sin6, sizeof(sin6)));
+          break;
+        }
+    }
+}
+#else
+#define _feed_aux_from_cmsg(aux, msg)
+#endif
+
+
+static gssize
+log_transport_udp_socket_read_method(LogTransport *s, gpointer buf, gsize buflen, LogTransportAuxData *aux)
+{
+  LogTransportUDP *self = (LogTransportUDP *) s;
+  gint rc;
+  struct msghdr msg;
+  struct iovec iov[1];
+  struct sockaddr_storage ss;
+#if defined(SYSLOG_NG_HAVE_CTRLBUF_IN_MSGHDR)
+  gchar ctlbuf[64];
+  msg.msg_control = ctlbuf;
+  msg.msg_controllen = sizeof(ctlbuf);
+#endif
+
+  msg.msg_name = (struct sockaddr *) &ss;
+  msg.msg_namelen = sizeof(ss);
+  msg.msg_iovlen = 1;
+  msg.msg_iov = iov;
+  iov[0].iov_base = buf;
+  iov[0].iov_len = buflen;
+
+  do
+    {
+      rc = recvmsg(self->super.super.fd, &msg, 0);
+    }
+  while (rc == -1 && errno == EINTR);
+
+  if (rc == 0)
+    {
+      /* DGRAM sockets should never return EOF, they just need to be read again */
+      rc = -1;
+      errno = EAGAIN;
+    }
+  else if (rc > 0)
+    {
+      if (msg.msg_namelen && aux)
+        log_transport_aux_data_set_peer_addr_ref(aux, g_sockaddr_new((struct sockaddr *) &ss, msg.msg_namelen));
+      _feed_aux_from_cmsg(self, aux, &msg);
+    }
+  return rc;
+
+}
+
+static void
+log_transport_udp_setup_fd(LogTransportUDP *self, gint fd)
+{
+  gint on = 1;
+  gint domain;
+  socklen_t domain_len = sizeof(domain);
+
+  self->local_addr = g_sockaddr_new_from_local_name(fd);
+
+  if (getsockopt(fd, SOL_SOCKET, SO_DOMAIN, &domain, &domain_len) < 0)
+    return;
+
+  if (domain == AF_INET)
+    {
+#ifdef __FreeBSD__
+      setsockopt(fd, IPPROTO_IP, IP_RECVDSTADDR, &on, sizeof(on));
+#else
+      setsockopt(fd, IPPROTO_IP, IP_PKTINFO, &on, sizeof(on));
+#endif
+    }
+  else if (domain == AF_INET6)
```

Comment:
`SYSLOG_NG_ENABLE_IPV6` is missing here?

Reply:
also done in 51f0fefab38fc4b18262547b6673db6be227be2b